### ISSUE TITLE
Update 721 note on transfers

### ISSUE THREAD

**Issue by fulldecent** [pixowl/thesandbox-contracts#1]
Hello! I saw this note in the code

```
// TODO should NFT be allowed to tranfer only if erc721_operators (or operatorsForAll) match ?
// in other word, do the ERC721 standard allow getApproved(tokenId) and isApprovedForAll(owner, operator) to return false 
// and at the same time still allow the transfer to pass (through mechanism defined outside the standard)?
```

I believe the question is: "may a valid implementation of ERC-721 perform allow a `transferFrom` to succeed if the caller is neither the current owner, an authorized operator, or the approved address for this NFT?"

The answer is no. It is specified (required) that a compliant ERC-721 implementation must have the transfer function throw if the caller is neither the current owner, an authorized operator, or the approved address for this NFT.

Reference: https://eips.ethereum.org/EIPS/eip-721 specifically "Every ERC-721 compliant contract must implement the ERC721... [interface] (subject to “caveats” below):" and documentation for `transferFrom`.

Conversely, it IS permissible for an ERC-721 contract's `transferFrom` to throw under additional circumstances than those required. And this is a key future compatibility / extendability feature.

**Comment by fulldecent** [comment]
To be clear. If your idea doesn't comply with a standard you can still do it, for sure! Just always understand the risks.

**Comment by wighawag** [comment]
Hey @fulldecent thanks for your comment!

rereading ERC721 it is indeed clear that

> the transfer function throw if the caller is neither the current owner, an authorized operator, or the approved address for this NFT.

This is unfortunate as the standard could have been more flexible in that regard easily.

There might be scope for interpretation though as an "authorized operator" could be meaning more than what the spec refers to. In other words should we interpret the words "authorized operator" as referring solely to what isApprovedForAll returns ?

If this is not the case and such interpretation is impossible, we might indeed be slightly out of compliance there since we want to allow our smart contract to be expandable via "super operators" that can be contract that include extra logic without requiring user to first approve them, for their convenience.

We could still be compliant if we make it isApprovedForAll  return true by default for such operators since the standard does not specify if these should be disabled by default or not.

Also do setApprovalForAll require to work for every input or can it throw for these operators for example (to forbid any user to disapprove them) ?

**Comment by fulldecent** [comment]
Yes! This is the type of creative thinking that gets things done!!!

Yes, by default, every blockchain function can throw if it wants to. This includes business reasons like you mentioned.
